native/dhseqr.go

```diff
@@ -0,0 +1,255 @@
+//      Aggressive Early Deflation. SIAM J. Matrix Anal. Appl. 23(4) (2002), pp. 948—973
+//      URL: http://dx.doi.org/10.1137/S0895479801384585
+//
+// Dhseqr is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dhseqr(job lapack.Job, compz lapack.Comp, n, ilo, ihi int, h []float64, ldh int, wr, wi []float64, z []float64, ldz int, work []float64, lwork int) (unconverged int) {
+	var wantt bool
+	switch job {
+	default:
+		panic(badJob)
+	case lapack.EigenvaluesOnly:
+	case lapack.EigenvaluesAndSchur:
+		wantt = true
+	}
+	var wantz bool
+	switch compz {
+	default:
+		panic("lapack: bad compz")
+	case lapack.None:
+	case lapack.InitZ, lapack.UpdateZ:
+		wantz = true
+	}
+	switch {
+	case n < 0:
+		panic(nLT0)
+	case ilo < 0 || max(0, n-1) < ilo:
+		panic(badIlo)
+	case ihi < min(ilo, n-1) || n <= ihi:
+		panic(badIhi)
+	case len(work) < lwork:
+		panic(shortWork)
+	case lwork < max(1, n) && lwork != -1:
+		panic(badWork)
+	}
+	if lwork != -1 {
+		checkMatrix(n, n, h, ldh)
+		switch {
+		case wantz:
+			checkMatrix(n, n, z, ldz)
+		case len(wr) < n:
+			panic("lapack: wr has insufficient length")
+		case len(wi) < n:
+			panic("lapack: wi has insufficient length")
+		}
+	}
+
+	const (
+		// Matrices of order ntiny or smaller must be processed by
+		// Dlahqr because of insufficient subdiagonal scratch space.
+		// This is a hard limit.
+		ntiny = 11
+
+		// nl is the size of a local workspace to help small matrices
+		// through a rare Dlahqr failure. nl > ntiny is required and
+		// nl <= nmin = Ilaenv(ispec=12,...) is recommended (the default
+		// value of nmin is 75). Using nl = 49 allows up to six
+		// simultaneous shifts and a 16×16 deflation window.
+		nl = 49
+	)
+
+	// Quick return if possible.
+	if n == 0 {
+		work[0] = 1
+		return 0
+	}
+
+	// Quick return in case of a workspace query.
+	impl.Dlaqr04(wantt, wantz, n, ilo, ihi, nil, 0, nil, nil, ilo, ihi, nil, 0, work, -1, 1)
+	lwkopt := max(max(1, n), int(work[0]))
+	if lwork == -1 {
+		work[0] = float64(lwkopt)
+		return 0
+	}
+
+	// Copy eigenvalues isolaged by Dgebal.
+	for i := 0; i < ilo; i++ {
+		wr[i] = h[i*ldh+i]
+		wi[i] = 0
+	}
+	for i := ihi + 1; i < n; i++ {
+		wr[i] = h[i*ldh+i]
+		wi[i] = 0
+	}
+
+	// Initialize Z to identity matrix if requested.
+	if compz == lapack.InitZ {
+		impl.Dlaset(blas.All, n, n, 0, 1, z, ldz)
+	}
+
+	// Quick return if possible.
+	if ilo == ihi {
+		wr[ilo] = h[ilo*ldh+ilo]
+		wi[ilo] = 0
+		return 0
+	}
+
+	// Dlahqr/Dlaqr04 crossover point.
+	nmin := impl.Ilaenv(12, "DHSEQR", string(job)+string(compz), n, ilo, ihi, lwork)
+	nmin = max(ntiny, nmin)
+
+	if n > nmin {
+		// Dlaqr0 for big matrices.
+		unconverged = impl.Dlaqr04(wantt, wantz, n, ilo, ihi, h, ldh, wr[:ihi+1], wi[:ihi+1],
+			ilo, ihi, z, ldz, work, lwork, 1)
+	} else {
+		// Dlahqr for small matrices.
+		unconverged = impl.Dlahqr(wantt, wantz, n, ilo, ihi, h, ldh, wr[:ihi+1], wi[:ihi+1],
+			ilo, ihi, z, ldz)
+		if unconverged > 0 {
```

Comment:
I hope that when `Dgeev` is finished, some test matrix (for instance, from Matrix Market) will actually trigger this failure. At the moment this branch is not covered by the tests.